### PR TITLE
[A2-784] UI http interceptor

### PR DESCRIPTION
**Pull request by msorens**
### :nut_and_bolt: Description

This sends selected projects along with every API request from the UI.

### :+1: Definition of Done

Projects appear in the HTTP request headers

image

### :athletic_shoe: Demo Script / Repro Steps

Select projects from the global projects filter then navigate somewhere in the UI (having self-refresh is in the browser is a separate card).

### :white_check_mark: Checklist

- [x] Necessary tests added/updated?
- [ ] Necessary docs added/updated?
- [x] Code actually executed?
- [x] Vetting performed (unit tests, lint, etc.)?

### CODE REVIEW

File: components/automate-gateway/gateway/middleware/auth_interceptor_internal_test.go [hunk 1]
```diff
@@ -79,6 +79,10 @@ func TestGetProjectsFromMetadata(t *testing.T) {
 		input    []string
 		expected []string
 	}{
+		"no projects": {
```
Comment (msorens) [inline components/automate-gateway/gateway/middleware/auth_interceptor_internal_test.go]: Drive-by

File: components/automate-ui/src/app/services/http/http-client-auth.interceptor.ts [hunk 2]
```diff
@@ -1,26 +1,44 @@
...
+  ) {
+    this.store.select(selectors.options)
+      .subscribe((options: ProjectsFilterOption[]) => {
+        this.projects = options.filter(p => p.checked).map(p => p.value).join(', ');
+      });
+  }
 
   intercept(request: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
+    let headers = request.headers.set('Authorization', `Bearer ${this.chefSession.id_token}`);
+    if (this.projects) {
```
Comment (msorens) [inline components/automate-ui/src/app/services/http/http-client-auth.interceptor.ts]: Important to only send when there are some; otherwise we are making an invalid request and get a 403, because all projects must have non-zero length.

File: components/automate-ui/src/app/services/projects-filter/projects-filter.service.ts [hunk 3]
```diff
@@ -9,17 +11,17 @@ const STORE_OPTIONS_KEY = 'projectsFilter.options';
 
 @Injectable()
 export class ProjectsFilterService {
```
Comment (msorens) [inline components/automate-ui/src/app/services/projects-filter/projects-filter.service.ts]: All drive-by: add some types.

File: components/automate-ui/src/app/services/http/http-client-auth.interceptor.ts [hunk 4]
```diff
@@ -1,26 +1,44 @@
...
-  constructor(private chefSession: ChefSessionService) {}
+  private projects: string;
+
+  constructor(
+    private chefSession: ChefSessionService,
+    private store: Store<NgrxStateAtom>
+  ) {
+    this.store.select(selectors.options)
+      .subscribe((options: ProjectsFilterOption[]) => {
+        this.projects = options.filter(p => p.checked).map(p => p.value).join(', ');
```
Comment (tylercloke) [inline components/automate-ui/src/app/services/http/http-client-auth.interceptor.ts]: Will this work if the store returns undefined?

**Comment by blakestier** [comment]
>Think we'll run into tricky business with the interceptor approach of injecting the project headers if we ever need to opt out of them, but we can cross that bridge when we get there :)

I think we're there-- the interceptor is adding the projects to the `introspect_projects` call (in lieu of empty `[]`)

We're still testing but we'll at least need some modifications

File: components/automate-ui/src/app/services/http/http-client-auth.interceptor.ts [hunk 5]
```diff
@@ -1,26 +1,44 @@
...
+      });
+  }
 
   intercept(request: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
+    let headers = request.headers.set('Authorization', `Bearer ${this.chefSession.id_token}`);
+    if (this.projects) {
+      headers = headers.set('projects', this.projects);
```
Comment (bcmdarroch) [inline components/automate-ui/src/app/services/http/http-client-auth.interceptor.ts]: one idea @scottopherson gave me: 
```
if (thisEndpointDoesntNeedOrSupportFilteringByProjects) {
      return next.handle(req.clone());
    }
```

we could write the helper `thisEndpointDoesntNeedOrSupportFilteringByProjects` and keep track of all the APIs that shouldn't be filtered inside it

**Comment by msorens** [comment]
@blakestier Latest commit addresses your changes.
My preference is that any code that can, should use the new library call I added, `ContextWithoutProjects`. It makes it very clear that a given block of code wants to operate unfettered by project filtering, and that knowledge is all localized to the code needing it.
It seems a lot cleaner than having to hard-code a list of URLs that the UI knows about for stripping projects from the header, not to mention that how are our users going to know that they should not provide a projects header for a given API call?

File: components/authz-service/server/v2/policy_refresher.go [hunk 6]
```diff
@@ -152,14 +152,8 @@ func (refresher *policyRefresher) RefreshAsync() error {
...
 func (refresher *policyRefresher) updateEngineStore(ctx context.Context) error {
-	// We need to remove project filters from the request context
-	// otherwise they will be applied for store updates.
-	// This will fail on service start context, so only remove projects if ok.
-	md, ok := metadata.FromIncomingContext(ctx)
-	if ok {
-		delete(md, "projects")
-		ctx = metadata.NewOutgoingContext(ctx, md)
-	}
+	// Engine updates need unfiltered access to all data.
```
Comment (srenatus) [inline components/authz-service/server/v2/policy_refresher.go]: This comment is a bit confusing to me... We're inside Authz-service, does it even matter if this is set or not? Might be missing something 💭

File: lib/grpc/auth_context/auth_context.go [hunk 7]
```diff
@@ -78,6 +78,18 @@ func NewOutgoingProjectsContext(ctx context.Context) context.Context {
 	return metadata.NewOutgoingContext(ctx, md)
 }
 
+// ContextWithoutProjects removes previously added projects from the GRPC metadata
+// for those system operations that must not be filtered by projects.
+func ContextWithoutProjects(ctx context.Context) context.Context {
```
Comment (srenatus) [inline lib/grpc/auth_context/auth_context.go]: One relevant detail is missing from comment and name: the translation from incoming to outgoing metadata. Can we fix this? 😉

File: components/authz-service/server/v2/policy_refresher.go (hunk 6 above)
Comment (blakestier) [reply components/authz-service/server/v2/policy_refresher.go]: We aren't using the `ctx` to filter projects within the `V2SetPolicies` call so we can remove all of this (not sure why we had it before 🤷‍♂ )

File: components/authz-service/server/v2/authz.go [hunk 8]
```diff
@@ -139,6 +140,10 @@ func (s *authzServer) FilterAuthorizedPairs(
 func (s *authzServer) FilterAuthorizedProjects(
 	ctx context.Context,
 	req *api.FilterAuthorizedPairsReq) (*api.FilterAuthorizedProjectsResp, error) {
+
+	// Introspection needs unfiltered access.
+	ctx = auth_context.ContextWithoutProjects(ctx)
```
Comment (blakestier) [inline components/authz-service/server/v2/authz.go]: Consider moving this to just after 155 since that's where we use it (I got confused and thought `V2FilterAuthorizedProjects`

File: components/authz-service/server/v2/policy_refresher.go [hunk 9]
```diff
@@ -152,14 +152,8 @@ func (refresher *policyRefresher) RefreshAsync() error {
...
+	ctx = auth_context.ContextWithoutProjects(ctx)
 
 	policyMap, err := refresher.getPolicyMap(ctx)
```
Comment (blakestier) [inline components/authz-service/server/v2/policy_refresher.go]: Hmm we might be doing it here . . .

Comment (blakestier) [reply components/authz-service/server/v2/policy_refresher.go]: We are! :bushes: :ignoreme:

File: components/authz-service/server/v2/policy_refresher.go (hunk 6 above)
Comment (msorens) [reply components/authz-service/server/v2/policy_refresher.go]: It is needed, because that method proceeds to get policies, roles, and rules from the store--and those would be filtered without this bit.

File: lib/grpc/auth_context/auth_context.go (hunk 7 above)
Comment (msorens) [reply lib/grpc/auth_context/auth_context.go]: Sure... but I am not sure on what is what; please suggest how to make it better!

File: components/authz-service/server/v2/policy_refresher.go (hunk 6 above)
Comment (srenatus) [reply components/authz-service/server/v2/policy_refresher.go]: ... but what's needed to avoid that is removing it from the incoming Metadata. Outgoing metadata is irrelevant here, since there's no grpc client calls involved in this code path, or is there?

File: lib/grpc/auth_context/auth_context.go (hunk 7 above)
Comment (srenatus) [reply lib/grpc/auth_context/auth_context.go]: See above, I wonder if this helper is a good idea as is. The only point where we need this incoming/outgoing translation is the gateway.

Comment (srenatus) [reply lib/grpc/auth_context/auth_context.go]: Well, it's got a use for sure. Just not inside authz-service. 😉

File: components/authz-service/server/v2/authz.go (hunk 8 above)
Comment (blakestier) [reply components/authz-service/server/v2/authz.go]: 🤔 Do we need this in `FilterAuthorizedProjects` at all?  I wonder if we really want to push it into `getAllowedProjects` (which calls `FilterAuthorizedProjects`) and is what `IntrospectAllProjects` calls.

File: components/automate-ui/src/app/services/http/http-client-auth.interceptor.ts (hunk 4 above)
Comment (scottopherson) [reply components/automate-ui/src/app/services/http/http-client-auth.interceptor.ts]: Stores have an initial value so they should never be undefined unless a reducer sets a value to be so. `selectors.options` starts off as an empty array.

File: components/automate-ui/src/app/services/http/http-client-auth.interceptor.ts (hunk 5 above)
Comment (scottopherson) [reply components/automate-ui/src/app/services/http/http-client-auth.interceptor.ts]: Didn't see this PR until this morning. I'd also recommend having separate `Auth` and `ProjectsFilter`  interceptors in the request chain, similar to the branch I was playing with last night:

https://github.com/chef/automate/compare/sc/projects-filter-data#diff-673f4abc0b622315151c256ba3238a8b
https://github.com/chef/automate/compare/sc/projects-filter-data#diff-b99ce0ca235bf0716d691eab97a28bf5R311

File: components/authz-service/server/v2/authz.go (hunk 8 above)
Comment (blakestier) [reply components/authz-service/server/v2/authz.go]: (or `IntrospectAllProjects` itself)

Comment (blakestier) [reply components/authz-service/server/v2/authz.go]: I guess it's all the same flow so 🤷‍♂ nvmd

File: components/automate-ui/src/app/services/http/http-client-auth.interceptor.ts (hunk 5 above)
Comment (scottopherson) [reply components/automate-ui/src/app/services/http/http-client-auth.interceptor.ts]:
```
@Injectable()
export class ProjectsFilterInterceptor implements HttpInterceptor {
  constructor(private projectsFilter: ProjectsFilterService) {}

  intercept(request: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {

    if (thisEndpointDoesntNeedOrSupportFilteringByProjects) {
      // pass the request through
      return next.handle(req.clone());
    }

    return this.projectsFilter.selectionValue$.pipe(
      first(),
      flatMap(selectionValue => {
        const headers = request.headers.set('projects', selectionValue);

        if (selectionValue === ProjectsConstants.ALL_RESOURCES_VALUE) {
          headers.delete('projects');
        }

        return next.handle(request.clone({ headers }));
      })
    );
  }
}
```
